### PR TITLE
raftstore-v2: add waterfall metrics

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?
Issue Number: Ref #12842 

What's Changed:

```commit-message
- add water metrics
- fix potential panic when destroying a peer
- fix incorrect store size
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore-v2/src/operation/command/mod.rs [hunk 1]
```diff
@@ -625,9 +669,11 @@ impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
         }
         control.need_flush = false;
         let flush_state = self.flush_state().clone();
-        if let Some(wb) = &mut self.write_batch && !wb.is_empty() {
+        if self.write_batch.as_ref().map_or(false, |wb| !wb.is_empty()) {
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/command/mod.rs:672]: I think the old version is more readable.

File: components/raftstore-v2/src/operation/command/mod.rs [hunk 2]
```diff
@@ -265,6 +289,26 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             .send(ApplyTask::CommittedEntries(apply));
     }
 
+    #[inline]
+    fn report_store_time_duration<T>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        entry_and_proposals: &mut [(Entry, Vec<CmdResChannel>)],
+    ) {
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/command/mod.rs:297]: missing checks like in `report_batch_wait_duration`

File: components/raftstore-v2/src/raft/apply.rs [hunk 3]
```diff
@@ -77,9 +80,12 @@ impl<EK: KvEngine, R> Apply<EK, R> {
         assert_ne!(applied_term, 0, "{}", SlogFormat(&logger));
         let applied_index = flush_state.applied_index();
         assert_ne!(applied_index, 0, "{}", SlogFormat(&logger));
+        let tablet = remote_tablet.latest().unwrap().clone();
+        let perf_context = tablet.get_perf_context(cfg.perf_level, PerfContextKind::RaftstoreApply);
         Apply {
             peer,
             tablet: remote_tablet.latest().unwrap().clone(),
```
Comment (tabokie) [inline components/raftstore-v2/src/raft/apply.rs:87]:
```suggestion
            tablet,
```

File: components/raftstore-v2/src/operation/command/mod.rs (hunk 2 above)
Comment (BusyJay) [reply components/raftstore-v2/src/operation/command/mod.rs:297]: It's on purpose. v1 doesn't check it either.

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 4]
```diff
@@ -317,6 +324,56 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
+    ///
+    /// The message is pushed into the send buffer, it may not be sent out until
+    /// transport is flushed explicitly.
+    fn send_raft_message_on_leader<T: Transport>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        msg: RaftMessage,
+    ) {
+        let message = msg.get_message();
+        if message.get_msg_type() == MessageType::MsgAppend
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply components/raftstore-v2/src/operation/ready/mod.rs:337]: It's not checked in v1.

**Review by tabokie — APPROVED** [review]

**Review by ethercflow — APPROVED** [review]
LGTM

File: components/raftstore-v2/src/operation/ready/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tonyxuqqi) [reply components/raftstore-v2/src/operation/ready/mod.rs:337]: > It's not checked in v1.

I saw there's some code path checking if the water fall metrics is enabled or not.

[the remark this answers is not in the document]
Comment (BusyJay) [reply components/raftstore-v2/src/operation/ready/mod.rs:337]: This is the v1 version, I don't see any check here.

https://github.com/tikv/tikv/blob/8e6e348505e7f1f7b5e023c00b30f90e8d1b4084/components/raftstore/src/store/peer.rs#L1815-L1835

**Pull request merged**